### PR TITLE
feat(forms): Adding dropdown category and related sub question type

### PR DESCRIPTION
**Pull request by ccailly**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 

Added `Dropdown` category and related sub-question type : `Dropdown`

Form editor view :
image
image

Form end user view :
image

### CODE REVIEW

File: src/Form/QuestionType/AbstractQuestionTypeSelectable.php [hunk 1]
```diff
@@ -137,21 +175,23 @@ public function getValues(?Question $question): array
...
+                {% set uuid = random() %}
+            {% endif %}
 
             <div
                 class="d-flex gap-1 align-items-center mb-2"
                 {{ extra_details ? 'data-glpi-form-editor-question-extra-details' : '' }}
             >
                 <i
+                    role="button"
+                    aria-label="{{ __('Move option') }}"
```
Comment (AdrienClairembault) [inline src/Form/QuestionType/AbstractQuestionTypeSelectable.php:194]: Translation in inline twig templates are not detected by our extraction tool.
We need to generate them in the PHP code and pass the value to the template.

File: src/Form/QuestionType/AbstractQuestionTypeSelectable.php [hunk 2]
```diff
@@ -163,6 +203,7 @@ class="ti ti-grip-horizontal cursor-grab ms-auto me-1"
                     name="default_value[]"
                     value="{{ uuid }}"
                     class="form-check-input" {{ checked ? 'checked' : '' }}
+                    aria-label="{{ __('Default option') }}"
```
Comment (AdrienClairembault) [inline src/Form/QuestionType/AbstractQuestionTypeSelectable.php:206]: Same here.

File: src/Form/QuestionType/AbstractQuestionTypeSelectable.php [hunk 3]
```diff
@@ -173,8 +214,11 @@ class="w-full"
                     name="options[{{ uuid }}]"
                     value="{{ value|e('html_attr') }}"
                     placeholder="{{ placeholder|e('html_attr') }}"
+                    aria-label="{{ __('Selectable option') }}"
```
Comment (AdrienClairembault) [inline src/Form/QuestionType/AbstractQuestionTypeSelectable.php:217]: Same here.

File: src/Form/QuestionType/AbstractQuestionTypeSelectable.php [hunk 4]
```diff
@@ -173,8 +214,11 @@ class="w-full"
...
                 >
                 <i
+                    role="button"
+                    aria-label="{{ __('Remove option') }}"
```
Comment (AdrienClairembault) [inline src/Form/QuestionType/AbstractQuestionTypeSelectable.php:221]: Same here.

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Two small bugs I've found:

### 1\) Wrong input name

image

Should be answers_141 since we switched to the new format.
This mean answers are not taken into account for this question type.

### 2\) Invalid order

image

image

It seems the dropdown is not refreshed after options order are modified using drag and drop.

The "old" order is still displayed until we do another action like changing the default value.

**Review by AdrienClairembault — APPROVED** [review]

File: src/Form/QuestionType/QuestionTypeDropdown.php [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    public function loadJavascriptFiles(): array
+    {
+        return array_merge(
+            parent::loadJavascriptFiles(),
+            ['js/form_question_dropdown.js']
+        );
+    }
+
+    #[Override]
+    public function getFooterScript(): string
```
Comment (cedric-anne) [inline src/Form/QuestionType/QuestionTypeDropdown.php:87]: I think it is not publicly used.
```suggestion
    protected function getFooterScript(): string
```

File: js/form_question_dropdown.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        super(inputType, container);
+
+        this._container.on('sortupdate', () => this.#handleSortableUpdate());
+
+        this._container.closest('[data-glpi-form-editor-question-details]')
+            .find('div[data-glpi-form-editor-specific-question-options]')
+            .find('input[data-glpi-form-editor-original-name=is_multiple_dropdown]')
+            .on('change', (event) => {
+                this._inputType = event.target.checked ? 'checkbox' : 'radio';
+                this.#updateInputType(this._inputType);
```
Comment (cedric-anne) [inline js/form_question_dropdown.js:54]: Cahnging the `_inputType` property should be done in the `#updateInputType()` method
```suggestion
                this.#updateInputType(event.target.checked ? 'checkbox' : 'radio');
```

File: js/form_question_dropdown.js [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            });
+    }
+
+    /**
+     * Update the input type of the options
+     *
+     * @param {string} inputType
+     */
+    #updateInputType(inputType) {
+        this._container.closest('[data-glpi-form-editor-question-type-specific]')
```
Comment (cedric-anne) [inline js/form_question_dropdown.js:87]:
```suggestion
        this._inputType = inputType;
        this._container.closest('[data-glpi-form-editor-question-type-specific]')
```

File: src/Form/QuestionType/AbstractQuestionTypeSelectable.php [hunk 8]
```diff
@@ -53,9 +53,41 @@ public function __construct()
...
+     * Some twig variables are available:
+     * - question: the question object
+     * - input_type: the input type
+     * - question_type: the question type class
+     * - rand: a random number
+     *
      * @return string
      */
-    abstract public function getInputType(): string;
+    public function getFooterScript(): string
```
Comment (cedric-anne) [inline src/Form/QuestionType/AbstractQuestionTypeSelectable.php:71]:
```suggestion
    protected function getFooterScript(): string
```

File: templates/pages/admin/form/form_question.html.twig [hunk 9]
```diff
@@ -46,9 +46,11 @@
     'no_label'     : true,
 } %}
 
-<div
+<section
+    role="option"
     class="mb-3"
     data-glpi-form-editor-question
+    aria-label="{{ question is not null ? question.fields.name|e('html_attr') : __('New question') }}"
```
Comment (cedric-anne) [inline templates/pages/admin/form/form_question.html.twig:53]: See #17147
```suggestion
    aria-label="{{ question is not null ? question.fields.name : __('New question') }}"
```

File: js/form_editor_controller.js [hunk 10]
```diff
@@ -451,8 +451,8 @@ class GlpiFormEditorController
 
             // Update input name
             let postfix = "";
-            if (typeof field === 'string' && field.match(/\[([\w[\]]*)\]$/g)) {
-                postfix = field.match(/\[([\w[\]]*)\]$/g);
+            if (typeof field === 'string' && field.match(/\[([\w-[\]]*)\]$/g)) {
+                postfix = field.match(/\[([\w-[\]]*)\]$/g);
```
Comment (cedric-anne) [inline js/form_editor_controller.js:455]: 1. Pattern could be placed in a constant.
2. The `RegExp.test()` method could be used in the `if` condition.
```suggestion
            cons postfix_pattern = new RegExp('\[([\w-\[\]]*)\]$', 'g');
            if (typeof field === 'string' && postfix_pattern.test(field)) {
                postfix = field.match(postfix_pattern);
```

**Comment by orthagh** [comment]
(opinion based on the provided screenshots)
I feel the design choice when checking multiple flag to switch between radio or checkboxes for options lead to a bad understanding. At least it took a few seconds to understand the options will be displayed in a real dropdown. I almost made a comment about why call it dropdown.

I think in this particular case of dropdown, you should not decorate options at all. More, there is the drag icon, so 2 icons on the left of the options

Ideas for alternative ux:

- select2 tokenisation (may not work without multiple select)
- modal with a [new option] link right to the dropdown
- keep the same design (without decoration), but make options list like it's the dropdown opened (check this [link](https://forum.smartflowsheet.com/support/solutions/articles/5000594816-how-do-i-add-delete-options-in-the-dropdown-picker-list-and-group-parameters-))

**Comment by ccailly** [comment]
Capture d’écran du 2024-05-30 11-29-49

**Comment by ccailly** [comment]
image

**Comment by cedric-anne** [comment]
@AdrienClairembault @orthagh
Is this OK for you now?

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
UI is good now.
Some changes requested for the code.

File: src/Glpi/Form/QuestionType/AbstractQuestionTypeActors.php [hunk 11]
```diff
@@ -119,7 +119,7 @@ public function getDefaultValue(?Question $question, bool $multiple = false): ar
     }
 
     #[Override]
-    public function renderAdministrationTemplate(?Question $question): string
+    public function renderAdministrationTemplate(?Question $question, array $params = []): string
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply src/Glpi/Form/QuestionType/AbstractQuestionTypeActors.php:122]: This way the relevant code is only stored in `AbstractQuestionTypeSelectable` and does not impact all questions for something that don't use.

**Review by AdrienClairembault — APPROVED** [review]

File: tests/cypress/e2e/form/question_types/dropdown.cy.js [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+        // Check selected options and option labels
+        checkSelectedOptions([0, 1], true);
+        checkOptionLabels(["Option 1", "Option 2", "Option 3"], true);
+
+        // Change the question type to "Dropdown (simple)"
+        cy.findByRole("checkbox", { name: "Allow multiple options" }).uncheck();
+
+        // Check selected options and option labels
+        checkSelectedOptions([1]);
+        checkOptionLabels(["Option 1", "Option 2", "Option 3"]);
```
Comment (cedric-anne) [inline tests/cypress/e2e/form/question_types/dropdown.cy.js:265]: Not a blocker, but, IMHO, when we switch from multiple to single value dropdonw, we should do this:
- if there are multiple values selected, all should be deselected,
- if there is a unique value selected, it should remain selected.

**Review by cedric-anne — COMMENTED** [review]
Seems globally OK, I just made a few minor remarks.

**Review by orthagh — APPROVED** [review]
As discussed together, the current design still has flaws (particularly choosing the default value dropdown, covering the rest of the design, leads to a strange state).
But they may be addressed later, especially with the right panel discussed many times.

Some I agree we're ok for this PR.

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**